keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,941 @@
+
+* Discovery merging:
+
+  - During upgrade or downgrade, it may be the case that no apiserver has a
+    complete list of available resources. To fix the problems mentioned, it's
+    necessary that discovery exactly matches the capability of the system. So,
+    we will use the storage version objects to reconstruct a merged discovery
+    document and serve that in all apiservers.
+
+Why so much work?
+* Note that merely serving 503s at the right times does not solve the problem,
+  for two reasons: controllers might get an incomplete discovery and therefore
+  not ask about all the correct resources; and when they get 503 responses,
+  although the controller can avoid doing something destructive, it also can't
+  make progress and is stuck for the duration of the upgrade.
+* Likewise proxying but not merging the discovery document, or merging the
+  discovery document but serving 503s instead of proxying, doesn't fix the
+  problem completely. We need both safety against destructive actions and the
+  ability for controllers to proceed and not block.
+
+### User Stories (Optional)
+
+#### Garbage Collector
+
+The garbage collector makes decisions about deleting objects when all
+referencing objects are deleted. A discovery gap / apiserver mismatch, as
+described above, could result in GC seeing a 404 and assuming an object has been
+deleted; this could result in it deleting a subsequent object that it should
+not.
+
+This proposal will cause the GC to see the complete list of resources in
+discovery, and when it requests specific objects, see either the correct object
+or get a 503 (which it handles safely).
+
+#### Namespace Lifecycle Controller
+
+This controller seeks to empty all objects from a namespace when it is deleted.
+Discovery failures cause NLC to be unable to tell if objects of a given resource
+are present in a namespace. It fails safe, meaning it refuses to delete the
+namespace until it can verify it is empty: this causes slowness deleting
+namespaces that is a common source of complaint.
+
+Additionally, if the NLC knows about a resource that the apiserver it is talking
+to does not, it may incorrectly get a 404, assume a collection is empty, and
+delete the namespace too early, leaving garbage behind in etcd. This is a
+correctness problem, the garbage will reappear if a namespace of the same name
+is recreated.
+
+This proposal addresses both problems.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Cluster admins might not read the release notes and realize they should enable
+network/firewall connectivity between apiservers. In this case clients will
+receive 503s instead of transparently being proxied. 503 is still safer than
+today's behavior.
```

Comment:
This can be further mitigated by documenting how to enable the feature to include a step for how to safely verify that the connectivity between apiservers is working as intended.

Reply:
we can also add a metric here to know of number of proxied connections succeed vs failed

Reply:
Added both suggested points.